sway/lock.c

```diff
@@ -0,0 +1,179 @@
+#define _POSIX_C_SOURCE 200809L
+#include "log.h"
+#include "sway/input/keyboard.h"
+#include "sway/input/seat.h"
+#include "sway/lock.h"
+#include "sway/output.h"
+#include "sway/server.h"
+
+struct sway_session_lock_surface {
+	struct wlr_session_lock_surface_v1 *lock_surface;
+	struct sway_output *output;
+	struct wlr_surface *surface;
+	struct wl_listener map;
+	struct wl_listener destroy;
+	struct wl_listener surface_commit;
+	struct wl_listener output_mode;
+};
+
+static void handle_surface_map(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, map);
+	sway_force_focus(surf->surface);
+	output_damage_whole(surf->output);
+}
+
+static void handle_surface_commit(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_surface* surf = wl_container_of(listener, surf, surface_commit);
+	output_damage_surface(surf->output, 0, 0, surf->surface, false);
+}
+
+static void handle_output_mode(struct wl_listener *listener, void *data) {
```

Comment:
The logical size of the output can change even if the mode doesn't, for instance when the scale or transform changes. We should listen for `output.events.commit` instead and check the `committed` bitfield in the event struct.

Reply:
I just tested using `output.events.commit`, and the signal (ignoring `committed == 3`) is not sent when a nested wayland display's window is resized.  It is correctly sent when using `swaymsg output` to adjust the mode, which I had not previously tested.  Is the lack of this signal for this a bug in wlroots, or do I need to hook both here?

Reply:
Hm right, to work correctly with current wlroots we probably need to listen to both.